source/slang/slang-emit-cuda.cpp

```diff
@@ -349,6 +349,150 @@ void CUDASourceEmitter::emitLoopControlDecorationImpl(IRLoopControlDecoration* d
     }
 }
 
+static bool _areEquivalent(IRType* a, IRType* b)
+{
+    if (a == b)
+    {
+        return true;
+    }
+    if (a->op != b->op)
+    {
+        return false;
+    }
+
+    switch (a->op)
+    {
+        case kIROp_VectorType:
+        {
+            IRVectorType* vecA = static_cast<IRVectorType*>(a);
+            IRVectorType* vecB = static_cast<IRVectorType*>(b);
+
+            if (GetIntVal(vecA->getElementCount()) != GetIntVal(vecB->getElementCount()))
+            {
+                return false;
+            }
+            return _areEquivalent(vecA->getElementType(), vecB->getElementType());
+        }
+        case kIROp_MatrixType:
+        {
+            IRMatrixType* matA = static_cast<IRMatrixType*>(a);
+            IRMatrixType* matB = static_cast<IRMatrixType*>(b);
+
+            if (GetIntVal(matA->getColumnCount()) != GetIntVal(matA->getColumnCount()) ||
+                GetIntVal(matA->getRowCount()) != GetIntVal(matA->getRowCount()))
+            {
+                return false;
+            }
+
+            return _areEquivalent(matA->getElementType(), matB->getElementType());
+        }
+        default:
+        {
+            IRBasicType* basicA = as<IRBasicType>(a);
+            IRBasicType* basicB = as<IRBasicType>(b);
+
+            return basicA && basicB;
+        }
+    }
+}
+
+void CUDASourceEmitter::_emitInitArrayValue(IRType* dstType, IRInst* value)
+{
+    // When constructing a matrix or vector from a single value this is already handled by the default path
+
+    switch (value->op)
+    {
+        case kIROp_Construct:
+        case kIROp_MakeMatrix:
+        case kIROp_makeVector:
+        {
+            IRType* type = value->getDataType();
+
+            // If the types are the same, we can can just break down and use
+            if (_areEquivalent(dstType, type))
+            {
+                if (auto vecType = as<IRVectorType>(type))
+                {
+                    if (UInt(GetIntVal(vecType->getElementCount())) == value->getOperandCount())
+                    {
+                        _emitInitArray(vecType->getElementType(), value->getOperands(), value->getOperandCount());
+                        return;
+                    }
+                }
+                else if (auto matType = as<IRMatrixType>(type))
+                {
```

Comment:
The logic in the matrix case is kind of complicated. Ideally we should be going back and vetting the IR to make sure things emit in a more predictable way:

* `makeVector` should always apply to `N` scalars and produce an `N`-component vector

* `makeMatrix` should always apply to `R` vectors (each with `C` components) and produce an `R` by `N` matrix

* `Construct` should be eliminated completely, and get replaced with more specific opcodes and/or explicit code in the stdlib.

(None of the above is a request for changes in this PR)

Reply:
Yes this would simplify things.